### PR TITLE
Port remaining benchmarks to BenchmarkDotNet

### PR DESCRIPTION
**Pull request by adamsitnik**

### CODE REVIEW

**Comment by adamsitnik** [comment]
@dotnet-bot test Innerloop Ubuntu16.04 Release Build and Test

File: tests/Benchmarks/Benchmarks.csproj [hunk 1]
```diff
@@ -9,6 +9,11 @@
     <Copyright>Microsoft Corporation, All rights reserved</Copyright>
     <NoWarn>$(NoWarn);CS0618</NoWarn>
   </PropertyGroup>
+  <ItemGroup>
+    <Compile Remove="BenchmarkDotNet.Artifacts\**" />
```
Comment (ahsonkhan) [inline tests/Benchmarks/Benchmarks.csproj]: What is the purpose of this ItemGroup?

File: tests/Benchmarks/Pipelines/E2EPipeline.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+using System.Text;
+using System.Text.Formatting;
+using System.Text.Json;
+using BenchmarkDotNet.Attributes;
+using BenchmarkDotNet.Attributes.Jobs;
+using BenchmarkDotNet.Engines;
+
+namespace Benchmarks.Pipelines
+{
+    [SimpleJob(RunStrategy.Monitoring)] // the setting for benchmarks with not-steady state
```
Comment (ahsonkhan) [inline tests/Benchmarks/Pipelines/E2EPipeline.cs]: Why is this necessary here? We previously ran it multiple times and reported the result as part of xunit.perf.

File: tests/Benchmarks/Program.cs [hunk 3]
```diff
@@ -17,9 +18,17 @@ public class Program
...
             => BenchmarkSwitcher
-                .FromAssembly(typeof(Program).Assembly)         
+                .FromAssemblyAndTypes(typeof(Program).Assembly, GetGenericBenchmarks())         
                 .Run(args/*, CreateClrVsCoreConfig() uncomment it to run Clr vs .NET Core comparison*/);
 
+        // BenchmarkDotNet can run generic benchmarks, if it knows what generic type arguments to use
+        private static Type[] GetGenericBenchmarks()
+            => new[]
+            {
+                typeof(HttpParser<Request>),
```
Comment (ahsonkhan) [inline tests/Benchmarks/Program.cs]: How do we extend this when new tests are added with different generic types?

File: tests/Benchmarks/System.Memory/IndexOf.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        static byte[] s_buffer = new byte[2000];
+
+        [GlobalSetup]
+        public void PrintInfo() => Console.WriteLine($"// Vector.IsHardwareAccelerated={Vector.IsHardwareAccelerated}"); // just print the info to help understand the results
+
+        [Benchmark(OperationsPerInvoke = 6 * 4)] // avoid loop overhead, do manuall loop unrolling
+        [Arguments(0)] // we don't have more arguments here because the CPU cache gets warmed-up very soon and the results are the same for every $at
+        [Arguments(1000)]
+        public int SpanIndexOf(int at)
```
Comment (ahsonkhan) [inline tests/Benchmarks/System.Memory/IndexOf.cs]: The previous test was measuring the IndexOf method and comparing the vectorized vs. non-vectorized, not the span indexer.

Also, do we have a way to programmatically set Vector.IsHardwareAccelerated to true/false? I know it can be done by setting an environment variable before running the test.

File: tests/Benchmarks/System.Memory/Slice.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
+﻿using BenchmarkDotNet.Attributes;
+using BenchmarkDotNet.Diagnosers;
+
+namespace System.Memory.Benchmarks
+{
+    /// <summary>
+    /// A class containing benchmarks for comparing constructs that do the same task in
+    /// different ways with the same end outcome.
+    /// </summary>
+    // [HardwareCounters(HardwareCounter.InstructionRetired)] // uncomment if you want to get hardware counters
```
Comment (ahsonkhan) [inline tests/Benchmarks/System.Memory/Slice.cs]: Can we un-comment this from the get go?

File: tests/Benchmarks/System.Memory/Slice.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
+﻿using BenchmarkDotNet.Attributes;
```
Comment (ahsonkhan) [inline tests/Benchmarks/System.Memory/Slice.cs]: Missing license header

File: tests/Benchmarks/Http/HttpParser.cs [hunk 7]
```diff
@@ -32,38 +31,33 @@ public class HttpParser
...
+public class Request : IHttpHeadersHandler, IHttpRequestLineHandler
 {
     //public Http.Method Method;
     //public Http.Version Version;
     //public string Path;
     //public string Query;
     //public string Target;
 
-    public Dictionary<string, string> Headers = new Dictionary<string, string>();
+    //public Dictionary<string, string> Headers = new Dictionary<string, string>();
```
Comment (ahsonkhan) [inline tests/Benchmarks/Http/HttpParser.cs]: Was commenting this out necessary for some benchmark related reason or is it just cleanup to be consistent with the rest of the commented out code?

File: tests/Benchmarks/Azure/CosmosDb.cs [hunk 8]
```diff
@@ -35,26 +35,26 @@ public class CosmosDb
     static Sha256 sha;
 
     [GlobalSetup]
-    public void CosmosDbBench()
+    public void Setup()
```
Comment (ahsonkhan) [inline tests/Benchmarks/Azure/CosmosDb.cs]: As part of this clean up change, can you add a System.Azure.Experimental.Benchmarks namespace and move this test to a System.Azure.Experimental directory under Benchmarks?

File: tests/Benchmarks/System.Text.Utf8/Utf8StringPerf.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
+﻿using System;
```
Comment (ahsonkhan) [inline tests/Benchmarks/System.Text.Utf8/Utf8StringPerf.cs]: license header

File: tests/Benchmarks/System.Text.Utf8/Utf8StringPerf.cs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            utf8String = new Utf8String(randomString);
+        }
+
+        [Benchmark]
+        public Utf8String ConstructFromString() => new Utf8String(randomString);
+
+        [Benchmark]
+        public void EnumerateCodePoints()
+        {
+            foreach (var codePoint in utf8String)
```
Comment (ahsonkhan) [inline tests/Benchmarks/System.Text.Utf8/Utf8StringPerf.cs]: nit: don't use var here

File: tests/Benchmarks/System.Text.Utf8/Utf8StringPerf.cs [hunk 11]
```diff
@@ -0,0 +1,49 @@
+﻿using System;
+using System.Text;
+using System.Text.Utf8;
+using BenchmarkDotNet.Attributes;
+
+namespace Benchmarks.System.Text.Utf8
```
Comment (ahsonkhan) [inline tests/Benchmarks/System.Text.Utf8/Utf8StringPerf.cs]: We decided on the namespace structure (like unit tests) as `<namespace>.Benchmarks`

File: tests/Benchmarks/System.Text.Utf8/Utf8StringPerf.cs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+{
+    public class Utf8StringPerf
+    {
+        [Params(5, 50000)]
+        public int Length;
+
+        [Params(126, 0xD7FF)]
+        public int MaxCodePoint;
```
Comment (ahsonkhan) [inline tests/Benchmarks/System.Text.Utf8/Utf8StringPerf.cs]: We end up losing the "description" of the input. "Short ASCII string", "Short string" etc. Any way we can get that back?

File: tests/Benchmarks/Benchmarks.csproj (hunk 1 above)
Comment (adamsitnik) [reply tests/Benchmarks/Benchmarks.csproj]: by default, VS displays the content of BenchmarkDotNet.Artifacts folder in the solution explorer. The folder contains results and logs. This changes is the result of right clicking on the folder and choosing "exclude from project"

File: tests/Benchmarks/Program.cs (hunk 3 above)
Comment (adamsitnik) [reply tests/Benchmarks/Program.cs]: In that case, we will need to specify the generic arguments like in this example.
Why is that? When BDN finds a generic type (let's say `Benchmark<T>`) it does not know which `T` should be used. So this has to be specified in an explicit way.

I just created an issue for that in BDN to allow doing this by using attributes: https://github.com/dotnet/BenchmarkDotNet/issues/745

File: tests/Benchmarks/System.Memory/Slice.cs (hunk 5 above)
Comment (adamsitnik) [reply tests/Benchmarks/System.Memory/Slice.cs]: sure, but please remember that to get Hardware Counters you need to run benchmarks as Admin (it's ETW limitation, BDN just warns about it)

File: tests/Benchmarks/Http/HttpParser.cs (hunk 7 above)
Comment (adamsitnik) [reply tests/Benchmarks/Http/HttpParser.cs]: I commented it because the corresponding struct type has this field commented too. I wanted to make them more "equal" to make the comparison easier.

File: tests/Benchmarks/System.Text.Utf8/Utf8StringPerf.cs (hunk 12 above)
Comment (adamsitnik) [reply tests/Benchmarks/System.Text.Utf8/Utf8StringPerf.cs]: Unfortunately not ;(

File: tests/Benchmarks/System.Memory/IndexOf.cs (hunk 4 above)
Comment (adamsitnik) [reply tests/Benchmarks/System.Memory/IndexOf.cs]: @ahsonkhan yes, it's possible. Could you tell me the name of env var which controlls that? I will add sample config then

**Comment by adamsitnik** [comment]
@ahsonkhan I pushed the fixes:

1. Adopted all types with benchmarks to the naming convention
2. Added missing license header
3. Fixed the IndexOf benchmark and added the config which allows comparing SIMD vs no SIMD

File: tests/Benchmarks/System.Memory/IndexOf.cs (hunk 4 above)
Comment (ahsonkhan) [reply tests/Benchmarks/System.Memory/IndexOf.cs]: You can disable vectorization by setting `COMPlus_FeatureSIMD=0`

File: tests/Benchmarks/System.Azure.Experimental/CosmosDb.cs [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+using System.Buffers.Text;
+using System.Globalization;
+using System.Security.Cryptography;
+using System.Text;
+using System.Text.Encodings.Web.Utf8;
+using System.Text.Utf8;
+using System.Runtime.InteropServices;
+using BenchmarkDotNet.Attributes;
+
+
```
Comment (ahsonkhan) [inline tests/Benchmarks/System.Azure.Experimental/CosmosDb.cs]: nit: extra new line

File: tests/Benchmarks/System.Memory/IndexOf.cs [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+    [Config(typeof(SimdConfig))]
+    public class IndexOf
+    {
+        static byte[] s_buffer = Enumerable.Repeat(byte.MinValue, 2000).ToArray();
+
+        [GlobalSetup]
+        public void PrintInfo() => Console.WriteLine($"// Vector.IsHardwareAccelerated={Vector.IsHardwareAccelerated}"); // just print the info to make sure the config works
+
+        [Benchmark(OperationsPerInvoke = 6 * 4)] // avoid loop overhead, do manuall loop unrolling
```
Comment (ahsonkhan) [inline tests/Benchmarks/System.Memory/IndexOf.cs]: Do we still need to do this and repeat the operation?

File: tests/Benchmarks/Pipelines/E2EPipeline.cs (hunk 2 above)
Comment (ahsonkhan) [reply tests/Benchmarks/Pipelines/E2EPipeline.cs]: Any thoughts on this, @adamsitnik?

File: tests/Benchmarks/System.Memory/IndexOf.cs (hunk 14 above)
Comment (adamsitnik) [reply tests/Benchmarks/System.Memory/IndexOf.cs]: @ahsonkhan yes, because the cost of creating Span is a big part of the benchmark for the cases where the searched element is at the beginning of memory buffer. 

``` ini

BenchmarkDotNet=v0.10.14.525-nightly, OS=Windows 10.0.16299.309 (1709/FallCreatorsUpdate/Redstone3)
Intel Xeon CPU E5-1650 v4 3.60GHz, 1 CPU, 12 logical and 6 physical cores
Frequency=3507503 Hz, Resolution=285.1031 ns, Timer=TSC
.NET Core SDK=2.1.300-rc1-008662
  [Host]        : .NET Core 2.1.0-rc1-26424-03 (CoreCLR 4.6.26424.05, CoreFX 4.6.26424.04), 64bit RyuJIT
  SIMD Disabled : .NET Core 2.1.0-rc1-26424-03 (CoreCLR 4.6.26424.05, CoreFX 4.6.26424.04), 64bit RyuJIT
  SIMD Enabled  : .NET Core 2.1.0-rc1-26424-03 (CoreCLR 4.6.26424.05, CoreFX 4.6.26424.04), 64bit RyuJIT

LaunchCount=1  TargetCount=3  WarmupCount=3  

```
|      Method |           Job |  EnvironmentVariables |   at |       Mean |      Error |    StdDev |
|------------ |-------------- |---------------------- |----- |-----------:|-----------:|----------:|
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** |    **0** |   **2.763 ns** |  **0.3031 ns** | **0.0171 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 |    0 |   5.815 ns |  2.5243 ns | 0.1426 ns |
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** |    **1** |   **3.086 ns** |  **0.9012 ns** | **0.0509 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 |    1 |   5.997 ns |  6.9594 ns | 0.3932 ns |
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** |    **2** |   **3.354 ns** |  **1.0120 ns** | **0.0572 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 |    2 |   5.776 ns |  2.7741 ns | 0.1567 ns |
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** |    **4** |   **3.280 ns** |  **0.4101 ns** | **0.0232 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 |    4 |   5.710 ns |  2.1890 ns | 0.1237 ns |
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** |    **8** |   **4.459 ns** |  **0.0775 ns** | **0.0044 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 |    8 |   5.941 ns |  0.8559 ns | 0.0484 ns |
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** |   **16** |   **5.976 ns** |  **1.8004 ns** | **0.1017 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 |   16 |   6.020 ns |  3.3568 ns | 0.1897 ns |
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** |   **30** |   **8.362 ns** |  **0.2031 ns** | **0.0115 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 |   30 |   6.405 ns |  3.3365 ns | 0.1885 ns |
| **SpanIndexOf** | **SIMD Disabled** | **COMPlus_FeatureSIMD=0** | **1000** | **225.725 ns** | **92.0383 ns** | **5.2003 ns** |
| SpanIndexOf |  SIMD Enabled | COMPlus_FeatureSIMD=1 | 1000 |  32.428 ns |  5.5044 ns | 0.3110 ns |

File: tests/Benchmarks/System.Memory/Slice.cs (hunk 5 above)
Comment (adamsitnik) [reply tests/Benchmarks/System.Memory/Slice.cs]: When running other benchmarks, it does not affect the experience.
When running this particular benchmark as non-admin, it's going to complain about it and run it without hardware counters.

So I think that we can keep it.

File: tests/Benchmarks/Pipelines/E2EPipeline.cs (hunk 2 above)
Comment (adamsitnik) [reply tests/Benchmarks/Pipelines/E2EPipeline.cs]: @ahsonkhan The default settings of BenchmarkDotNet use some heuristics to determine how many times to run the benchmark to get stable results. This is perfect for steady-state benchmarks. 

However, benchmarks which don't have a steady state (like heavy IO benchmarks) should not use the default settings. Their distribution is typically very different, which does not work well with our heuristics.

[Monitoring](https://benchmarkdotnet.org/Guides/ChoosingRunStrategy.htm) strategy runs it a certain amount of times, does not pre-warmup the code and does not exclude the outliers. It's preferred strategy for benchmarks like this one.
